Alternative scope selector

leio10 opened the pull request:
#### :tophat: What? Why?
Current scope selector based on Select2 works better for a large amount of scopes than a standard HTML select. But it has, at least, two problems:
 - It is not very responsive. Using it with mobile phones is hard, since Select2 doesn't seems to be responsive. Using the mobile on-screen keyword reduces the browser area used to display options.
 - It flattens the underlying hierarchical structure of scopes. The user does not have a clear view of this structure to allow him, for example, to choose a neighborhood instead of a district.

After some research about alternative inputs for this case, I've found [this article](
https://medium.com/@mibosc/responsive-design-why-and-how-we-ditched-the-good-old-select-element-bc190d62eff5) that presents a solution that I liked very much. As the article says, this solution could be used for many cases, so I implemented a generic data picker, and then configured an specific scopes picker. That means that it should not be hard to implement new pickers based on this code.

#### :pushpin: Related Issues
- Related to #1747

#### :clipboard: Subtasks
- [x] Implement data picker
- [x] Implement scope picker
- [x] Replace all scope selectors with scope pickers
- [x] Implement multiple data picker
- [x] Replace multiple scopes filter with multiple data picker
- [x] Support for History browsing in filters component
- [x] Remove scopes selector (?)
- [x] Remove select2 (?)
- [x] Complete tests

### :camera: Screenshots (optional)
![picker_admin](https://user-images.githubusercontent.com/453545/33612598-5ea2e358-d9d2-11e7-81c4-adfa78092240.gif)
![picker_admin_movil](https://user-images.githubusercontent.com/453545/33613239-62c42116-d9d4-11e7-91e5-8abfb472c62d.gif)

#### :ghost: GIF
![]()

xabier commented:
Looks great!

josepjaume commented:
This is ultra nice!!! ❤️ ❤️ 

I feel like we're missing a search box though. Might be useful to search across any level of hierarchy, right?

In any case, extra good job!

leio10 commented:
@josepjaume @xabier thanks for the feedback!

I've implemented multiple scopes selector, trying to keep it as simple as I can. This is what I got:
![picker_proposals_filter](https://user-images.githubusercontent.com/453545/33833451-f99be434-de7f-11e7-9e49-f52f45f058b5.gif)

Now, I'm going to change select2 with this in all components and then fix and add missing tests.
Should I remove select2 completely? Do you know if it is used anywhere else?

@josepjaume Regarding searches, I think that browsing scopes hierarchically make the searching feature not mandatory. So, I prefer to add it in a later PR, thinking before what scopes to include in the search (can be root scope descendants, current scope descendants or only current scope children) and how to show it on mobile devices. Are you okay with this?

mrcasals commented:
@leio10 grepping the source code it looks like `select2` is only used for scopes forms. It's being used in `decidim-admin` and `decidim-proposals` mostly, but some code lies in `decidim-core` to avoid duplication. I'm 👍  on removing it!

Also, @deivid-rodriguez will be happy as it took him quite a few PRs to fix some weird random tests 😁

leio10 commented:
Ok, as soon as tests are finished this should be ready to be merged.

@josepjaume I didn't add the search feature yet, please tell me if you think is mandatory to merge this change.

@deivid-rodriguez Sorry for all the extra work generated by Select2, it seems a good idea at that moment... :speak_no_evil:

Finally, I would like three changes that I think that you should know about it:

- https://github.com/decidim/decidim/pull/2330/commits/d35058db1ac7671c7373aca86a85beda8cb7902f: I modified History object to allow to save a state object within URLs. I use this to store and retrieve picker selected values (scope ids, labels and link URL). Doing this without History's state would very complicated.
- https://github.com/decidim/decidim/pull/2330/commits/4a84d269a86b3197c55bcf2bb3f605befa9e70b2: I modified some forms in several modules (proposals admin and front, budgets, meetings and results) to use participatory space scope_id when none is selected. Both four forms do the same things (expose scope_id, scope and categories), but all with different code. This change is mostly not required for this PR, but using the exactly same code in all of them makes easy to avoid bugs. Also, maybe we should consider to move this to concerns or something similar?
- https://github.com/decidim/decidim/pull/2330/commits/b536035560cde8629ed10879dceefd5d90219cbd: I made public the `label_for` method at `FormBuilder` class. I use this to the build picker title in forms ([here](https://github.com/decidim/decidim/pull/2330/commits/18f4d24b51d084d9eff25364ca80b28eb1b95df9), [here](https://github.com/decidim/decidim/pull/2330/commits/11f1187c6c611d90d89330b2ddef50d0a69248be) and [here](https://github.com/decidim/decidim/pull/2330/commits/80c71e066106dfaae72953db6532f4ae8a131ca8)).

mrcasals commented on decidim-accountability/app/forms/decidim/accountability/admin/result_form.rb:
```diff
@@ -38,12 +38,25 @@ def proposals
 
-        def organization_scopes
-          current_organization.scopes
+        def participatory_space_scope
```
We have the `HasScope` concern, but would not apply here since those are form objects, and the already existing concern is for models. 

I'm 👍  on using a concern for this, but we have to be careful with the naming, and have some good docs warning developers this concern should only be used for form objects.

leio10 replied:
Well, I'm not sure if this should be a concern. All that forms have a scope and a category, and updating all every time you change one is tedious, but maybe creating concerns in the core engine for forms defined in modules increase complexity and reduce flexilibity. :thinking:

deivid-rodriguez replied:
Mmmmm what about moving these common methods up to `Decidim::Form`? Or maybe better, create a `Decidim::FormOnFeature` equivalent to `Decidim::Form` that assumes the availability of `current_feature` and inherit from that instead? Are those checks on `current_feature` actually needed or these forms always need a `current_feature`?

mrcasals commented on decidim-proposals/app/views/decidim/proposals/admin/proposals/_form.html.erb:
```diff
@@ -26,7 +26,10 @@
       <div class="row column">
-        <%= form.scopes_select :scope_id, prompt: I18n.t("decidim.scopes.global"), remote_path: decidim.scopes_search_path(root: current_participatory_space.scope) %>
+        <%= form.scopes_picker :scope_id do |scope|
+          { url: decidim.scopes_picker_path(root: current_participatory_space.scope, current: scope&.id, field: form.label_for(:scope_id)),
```
I see this repeated in other files, what do you think about moving this logic to a helper method?

josepjaume replied:
It definitely looks like it should go to a helper in core!

leio10 replied:
Yes, you are totally right, I have had to make multiple replacements many times and I didn't see it :disappointed:.

leio10 commented on decidim-admin/app/assets/stylesheets/decidim/admin/modules/_data-picker.scss:
```diff
@@ -0,0 +1,111 @@
+.data-picker {
```
That would be great, but how do you do that? I searched for an example in the project and didn't find it. Even, I found the decidim/-admin/app/assets/stylesheets/decidim/admin/utils that has a lot of files copied from decidim-core/app/assets/stylesheets/decidim/utils.

mrcasals replied:
Mhhhh, maybe @josepjaume can help here?

josepjaume replied:
@leio10 the admin interface is borrowing most of the files but we probably shouldn't be doing that anymore. Have you tried just:

```scss
@import "decidim/modules/data_picker"';
```

Thanks!

leio10 replied:
Mmm, it doesn't work. I did a little research and I'm afraid that it is more complicated than that, and probably implies changes in the decidim-core gem. For example, https://github.com/zurb/foundation-rails includes the shaded SCSS files in the vendor folder, I don't know if that is necesary.

deivid-rodriguez replied:
@leio10 Can you try:

```diff
diff --git a/decidim-admin/app/assets/stylesheets/decidim/admin/_decidim.scss b/decidim-admin/app/assets/stylesheets/decidim/admin/_decidim.scss
index d17d5e76a..eb597239b 100755
--- a/decidim-admin/app/assets/stylesheets/decidim/admin/_decidim.scss
+++ b/decidim-admin/app/assets/stylesheets/decidim/admin/_decidim.scss
@@ -12,4 +12,6 @@
 
 @import "plugins/foundation-datepicker";
 
+@import "decidim/modules/data-picker";
+
 @import "modules/modules";
diff --git a/decidim-admin/app/assets/stylesheets/decidim/admin/modules/_data-picker.scss b/decidim-admin/app/assets/stylesheets/decidim/admin/modules/_data-picker.scss
deleted file mode 100644
index c5cc0e336..000000000
--- a/decidim-admin/app/assets/stylesheets/decidim/admin/modules/_data-picker.scss
+++ /dev/null
@@ -1,143 +0,0 @@
-.data-picker {
-  &.picker-single .picker-values div a, .picker-prompt a {
-    background: white;
-    cursor: pointer;
-    display: block;
-    width: 100%;
-    margin-bottom: 1rem;
-    padding: 0.7rem 0.5rem;
-    border: 1px solid #eee;
-    border-radius: 4px;
-    box-shadow: inset 0 1px 2px rgba(26, 24, 29, 0.1);
-    -webkit-box-shadow: inset 0 1px 2px rgba(26, 24, 29, 0.1);
-    color: #1a181d;
-
-    &:active {
-      outline: none;
-      border: 1px solid #1a181d;
-      background-color: #fff;
-      box-shadow: 0 0 5px #adadad;
-      -webkit-box-shadow: 0 0 5px #adadad;
-      transition: border-color 0.25s ease-in-out, -webkit-box-shadow 0.5s;
-      -webkit-transition: border-color 0.25s ease-in-out, -webkit-box-shadow 0.5s;
-    }
-
-    &::after {
-      display: block;
-      width: 0;
-      height: 0;
-      border: inset 5px;
-      content: '';
-      border-right-width: 0;
-      border-color: transparent transparent transparent #1a181d;
-      float: right;
-      margin-top: 7px;
-    }
-  }
-
-  &.picker-single .picker-values:not(:empty) + .picker-prompt {
-    display: none;
-  }
-
-  &.disabled.picker-single .picker-values div a, &.disabled .picker-prompt a {
-    cursor: not-allowed;
-    background: #eee;
-
-    &:active {
-      border: inherit !important;
-      box-shadow: inherit !important;
-      -webkit-box-shadow: inherit !important;
-      transition: none !important;
-      -webkit-transition: none !important;
-    }
-  }
-
-  &.picker-multiple .picker-values div {
-    display: block;
-    margin: 0 0 2px;
-    font-size: 0.9rem;
-
-    input {
-      margin: 2px 8px 8px 2px;
-      vertical-align: top;
-    }
-  }
-}
-
-#data_picker-modal {
-  height: 90%;
-
-  .data_picker-modal-content {
-    display: flex;
-    flex-direction: column;
-    max-height: 100%;
-    min-height: 100%;
-
-    .picker-footer {
-      .buttons {
-        display: flex;
-        justify-content: center;
-        padding: 10px 0px 50px; // bottom padding is needed to avoid hiding buttons on mobiles because of https://stackoverflow.com/questions/37112218/css3-100vh-not-constant-in-mobile-browser
-      }
-    }
-
-    .picker-content {
-      flex-grow: 10;
-      min-height: 100%;
-      max-height: 100%;
-      overflow: auto;
-
-      .picker-scrollable-content {
-        max-height: 100%;
-        overflow: auto;
-      }
-    }
-  }
-}
-
-.scope-picker {
-  ul {
-    margin-left: 0;
-    li {
-      list-style: none;
-    }
-  }
-
-  &.picker-header {
-    padding: 5px;
-    li:hover {
-      background: #ddd;
-    }
-
-    li ~ li a:before {
-      content: '';
-      width: 10px;
-      height: 2em;
-      float: left;
-    }
-    li:last-child a {
-      color: #333;
-    }
-    a {
-      display: block;
-      padding: 2px;
-    }
-  }
-
-  &.picker-content {
-    .picker-scrollable-content {
-      scroll-snap-type: mandatory;
-
-      a {
-        scroll-snap-coordinate: 0px 0px;
-        display: block;
-        border-top: 1px solid #ddd;
-        line-height: 1.5rem;
-        padding: 5px;
-        &:hover {
-          background: #ddd;
-        }
-      }
-    }
-  }
-}
\ No newline at end of file
diff --git a/decidim-admin/app/assets/stylesheets/decidim/admin/modules/_modules.scss b/decidim-admin/app/assets/stylesheets/decidim/admin/modules/_modules.scss
index 29f6d9fbc..2b9ebc4b3 100644
--- a/decidim-admin/app/assets/stylesheets/decidim/admin/modules/_modules.scss
+++ b/decidim-admin/app/assets/stylesheets/decidim/admin/modules/_modules.scss
@@ -15,9 +15,6 @@
 //Process
 @import "process-header";
 
-//Shared modules
-@import "data-picker";
-
 //Contents
 @import "callouts";
 @import "cards";
diff --git a/decidim-core/lib/decidim/core/engine.rb b/decidim-core/lib/decidim/core/engine.rb
index 206dd1d87..c48dd0c16 100644
--- a/decidim-core/lib/decidim/core/engine.rb
+++ b/decidim-core/lib/decidim/core/engine.rb
@@ -56,6 +56,8 @@ module Decidim
       end
 
       initializer "decidim.assets" do |app|
+        app.config.assets.paths << File.expand_path("../../../app/assets/stylesheets", __dir__)
+
         app.config.assets.precompile += %w(decidim_core_manifest.js)
 
         Decidim.feature_manifests.each do |feature|
```

leio10 replied:
Yeah! Works like a charm :+1:
I'll add it now and I hope that adding a new assets path don't change anything else. I suppose that this is the right way to do it, so we will should start to remove duplicated files, as in utils folder.

deivid-rodriguez replied:
Yeah, I have the same doubts / expectations / suppositions... Let's hope it doesn't break anything and we don't find any really bad side effects from doing this.

leio10 replied:
I guess that it should affects only to the admin appareance. I browsed it and checked that it seems to be the same. Anyway, I think that maybe @josepjaume has something to say :smiley_cat:

deivid-rodriguez replied:
I was thinking if it'd be better to go with a more conservative approach of only explicitly adding the specific sheets core wants to make public instead of sharing the whole design?

leio10 commented:
I've created the helpers and code is much nice now. But, I found a little problem with translations: moving them to a helper make them lose context (`.scope_prompt` and `.scopes`). Those translations are always the same texts, and were present in many modules (budgets, meetings, proposals, ...).

So, I moved all those translations to decidim-core, (also `decidim.scopes.global` that was in participatory-processes). BTW, I only moved english translations and leave others as they are, waiting to crowdin to fix it. Are you ok with this?

deivid-rodriguez commented:
@mrcasals Yes, it's a big pretty (that looks amazing). I'll try to give it a review later!

mrcasals commented:
@leio10 to me, the translations for "scopes" should be consistent all across the app, so keeping them in `core` works for me. And yes, only modify English translations, Crowdin will take care of the rest!

Will try to re-review this PR later on 😄

deivid-rodriguez commented on decidim-surveys/spec/queries/decidim/surveys/survey_user_answers_spec.rb:
```diff
@@ -14,6 +14,7 @@
     result = subject.query
+    result.reverse! unless result.first.first.user == user_1 # ensures that the first record is the first user
 
     expect(result[0]).to eq(survey_answers_user_1.sort { |answer| answer.question.position })
```
@leio10 For these things, I usually do something like `expect(result).to contain_exactly(survey_answers_user_1, survey_answers_user_2)`. Would that work here?

leio10 replied:
I'm not totally sure. I don't like it either but, according to the `sort` calls below, I guess that the answers could also be retrieved with a different order each time. 

Maybe we can do this
```ruby
result.reverse! unless result.first.first.user == user_1
expect(result[0]).to contain_exactly(survey_answers_user_1)
expect(result[1]).to contain_exactly(survey_answers_user_2)
```
but I can't figure out a simpler solution.

deivid-rodriguez replied:
But I think that's what `contain_exactly` do, treat the array as a set so that order doesn't matter...

leio10 replied:
Mmm, and it would work with an array of arrays? 
In other words, this would work?
```ruby
expect([[3, 2, 1], [6, 5, 4]]).to contain_exactly([1, 2, 3], [4, 5, 6])
```
My bet is no, because `contain_exactly` doesn't require that received params are arrays. Anyway, I will check it now... :smiley:

deivid-rodriguez replied:
And if you keep the ordering that's already there?

```
expect(result).to contain_exactly(survey_answers_user_1.sort { |answer| answer.question.position }, survey_answers_user_2.sort { |answer| answer.question.position })
```

 The query seems to order the answers by position...

leio10 replied:
You are right, I think that should work always :+1:

mrcasals replied:
Ther's also `match_array`, but I'm not sure it fits what you need

deivid-rodriguez commented:
By the way @leio10, this is some amazing piece of work :heart:

I don't have time to review the code in depth, but I tested the feature as a user this morning and it looks just great. I couldn't find any bugs or gotchas either, it just did what I expected it to do.

BTW: Loving also that you're not giving up and diving into flaky haunting as they show up! :muscle:

leio10 commented:
thanks @deivid-rodriguez! :heart: :heart: :heart: :heart: 
it's such a pleasure to learn from you and the rest of the team in this project :heart_eyes:

leio10 commented:
Well, I reviewed resources forms again and:
- enforced coherence between forms code (removed use of `context` to access `current_feature`, only when I sure: removed some useless aliases, removed some useless checks)
- replaced `where(...).first` with `find_by`
- show field errors in data-pickers and remove errors when value is changed
- don't allow scopes picker to select a scope that is not descendant from root scope
- added a helper to Form that allows to access to participatory space scope, if exists
- checked that selected scope is a descendant of participatory space scope (this is to avoid user attacks or if its participatory space scope is changed after the proposal is created).
- added a few tests related
![picker_errors](https://user-images.githubusercontent.com/453545/34216783-a07125be-e5a9-11e7-9437-92dd9a5d1f61.gif)

I think that with this change this feature its pretty complete. Also, we get rid of some flaky tests from select2 and, hopefully, a few more flaky tests.

So, @mrcasals @josepjaume, do you think that it is anything else that I can do to ease this review? will you have a while to review it? :smile_cat:

josepjaume commented on decidim-core/lib/decidim/form_builder.rb:
```diff
@@ -131,29 +131,43 @@ def categories_select(name, collection, options = {})
-      remote_path = options.delete(:remote_path) || false
-      multiple = options.delete(:multiple) || false
-      html_options = { multiple: multiple, class: "select2", "data-remote-path" => remote_path, "data-placeholder" => prompt }
+    def scopes_picker(attribute, options = {})
```
This is fine like this - but when the complexity of these things arises and is mostly markup, what I usually do is extract the code into a partial somewhere and just render that. This way you can leverage the clearness of the markup in a partial and have the logic be passed via locals (and prepopulated before).

It's something to consider the next time we step in this code 👍

josepjaume commented:
Merging straight in!!

deivid-rodriguez commented:
Yeah, this PR was totally EPIC. Outstanding work made here! :100:
